[StructuredAuthenticationConfig] Create struct for authn config and re-wire OIDC flags to use it

aramase opened the pull request:
- Adds the `v1alpha1` API for `AuthenticationConfiguration`
- Wires up existing `--oidc-*` flags with the internal struct
  - The `AuthenticationConfiguration` validation is used after the struct is created.

Fixes https://github.com/kubernetes/kubernetes/issues/118832
Fixes https://github.com/kubernetes/kubernetes/issues/119384

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3331-structured-authentication-configuration
```

aramase committed:
add AuthenticationConfiguration v1alpha1 api

Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

aramase committed:
wiring existing oidc flags with internal API struct

Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

aramase commented:
/kind feature
/sig auth
/triage accepted
/milestone v1.28

aramase commented:
/assign @enj 

Not ready for review yet, but just adding it to your list!

furkatgofurov7 commented:
@aramase hey, Bug Triage Lead for 1.28 release here 👋🏼 

I'd like to check what's the status of this PR since the code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about 3 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.28, is it still planned for this release?

aramase commented:
/priority important-soon

aramase commented:
> As the PR is tagged for 1.28, is it still planned for this release?

Yes, it's planned for v1.28 release. I've tagged the PR with priority.

aramase commented:
/cc @liggitt

furkatgofurov7 commented:
@aramase 👋🏼

This is a reminder, as I informed about it already 2 weeks ago, we are in week 9 currently and the code freeze is starting next week (week 10) at [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496).

Although it does not have any reviews yet (@liggitt @enj folks PTAL when you have some time), PR looks very close to the finish line. Please make sure that it is on track and have sufficient labels in order to make it to release 1.28 before code freeze is in. Also, let me know if you need any help/assistance.

enj commented on pkg/kubeapiserver/options/authentication.go:
```diff
@@ -397,16 +400,51 @@ func (o *BuiltInAuthenticationOptions) ToAuthenticationConfig() (kubeauthenticat
+		authConfig := &authenticationapi.AuthenticationConfiguration{
+			JWT: []authenticationapi.JWTAuthenticator{jwtAuthenticator},
+		}
+		if fieldErrList := authenticationapivalidation.ValidateAuthenticationConfiguration(authConfig); fieldErrList.ToAggregate() != nil {
+			return kubeauthenticator.Config{}, fieldErrList.ToAggregate()
```
```suggestion
		if err := authenticationapivalidation.ValidateAuthenticationConfiguration(authConfig).ToAggregate(); err != nil {
			return kubeauthenticator.Config{}, err
```

enj commented on pkg/kubeapiserver/options/authentication.go:
```diff
@@ -397,16 +400,51 @@ func (o *BuiltInAuthenticationOptions) ToAuthenticationConfig() (kubeauthenticat
+
+		authConfig := &authenticationapi.AuthenticationConfiguration{
+			JWT: []authenticationapi.JWTAuthenticator{jwtAuthenticator},
+		}
+		if fieldErrList := authenticationapivalidation.ValidateAuthenticationConfiguration(authConfig); fieldErrList.ToAggregate() != nil {
```
I am not sure if this is the correct place to call the validation logic, normally options are validate right after parsing in a `Validate` function.

aramase replied:
As discussed last week, I've left the validation here without any changes. LMK what you think.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/doc.go:
```diff
@@ -0,0 +1,20 @@
+
+// +k8s:deepcopy-gen=package
+
+// package config contains PodSecurity admission configuration file types
```
Same copy paste error.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/scheme/scheme.go:
```diff
@@ -0,0 +1,44 @@
+	AddToScheme(Scheme)
+}
+
+// AddToScheme builds the podsecurity scheme using all known versions of the podsecurity api.
```
Copy paste error.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/scheme/scheme.go:
```diff
@@ -0,0 +1,44 @@
+)
+
+var (
+	// Scheme is the runtime.Scheme to which all podsecurity api types are registered.
```
Copy paste error.

enj commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -302,35 +260,30 @@ func New(opts Options) (*Authenticator, error) {
 
 	verifierConfig := &oidc.Config{
-		ClientID:             opts.ClientID,
+		ClientID:             opts.JWTAuthenticator.Issuer.ClientIDs[0],
```
Validate that the length of this slice is equal to `1`.

aramase replied:
Done in validation!

enj commented on pkg/kubeapiserver/options/authentication.go:
```diff
@@ -397,16 +400,51 @@ func (o *BuiltInAuthenticationOptions) ToAuthenticationConfig() (kubeauthenticat
 	}
 
-	if o.OIDC != nil {
-		ret.OIDCCAFile = o.OIDC.CAFile
```
I am wondering if we can update `OIDCAuthenticationOptions` to contain a `JWTAuthenticator` and then validate is here:

https://github.com/kubernetes/kubernetes/blob/1db4658614023c3630a2e8d9a634eec4980d074f/pkg/kubeapiserver/options/authentication.go#L186-L191

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/v1alpha1/types.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+type JWTAuthenticator struct {
+	// issuer is a basic OIDC provider connection options.
```
```suggestion
	// issuer contains the basic OIDC provider connection options.
```

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/v1alpha1/types.go:
```diff
@@ -0,0 +1,109 @@
+	// At least one of the entries must match the "aud" claim in presented JWTs.
+	// Same value as the --oidc-client-id flag (though this field supports an array).
+	// Required to be non-empty.
+	ClientIDs []string `json:"clientIDs"`
```
Just a thought that came up as I was reading this: maybe we should rename this field to `Audiences`.

aramase replied:
opened https://github.com/kubernetes/kubernetes/issues/119384 to track this change

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/v1alpha1/types.go:
```diff
@@ -0,0 +1,109 @@
+type AuthenticationConfiguration struct {
+	metav1.TypeMeta
+
+	// jwt is a list of OIDC providers to authenticate Kubernetes users.
```
I think we need to come up with a way to describe this without using OIDC since we really only use that for public key discovery.

aramase replied:
Use text from https://kubernetes.io/docs/reference/access-authn-authz/authentication/#openid-connect-tokens and updated! LMK what you think.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/v1alpha1/types.go:
```diff
@@ -0,0 +1,109 @@
+	// claim is the JWT claim to use.
+	// +optional
+	Claim string `json:"claim"`
+	// prefix is prepended to claim to prevent clashes with existing names.
```
```suggestion
	// prefix is prepended to claim's value to prevent clashes with existing names.
```

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/v1alpha1/types.go:
```diff
@@ -0,0 +1,109 @@
+	// (1) and (2) ensure backward compatibility with the --oidc-username-claim and --oidc-username-prefix flags
+	Username PrefixedClaimOrExpression `json:"username"`
+	// groups represents an option for the groups attribute.
+	// Claim must be a string or string array claim.
```
```suggestion
	// The claim's value must be a string or string array claim.
```

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/v1alpha1/types.go:
```diff
@@ -0,0 +1,109 @@
+type ClaimValidationRule struct {
+	// claim is the name of a required claim.
+	// Same as --oidc-required-claim flag.
+	// Only string claims are supported.
```
```suggestion
	// Only string claim keys are supported.
```

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/v1alpha1/types.go:
```diff
@@ -0,0 +1,109 @@
+
+type ClaimMappings struct {
+	// username represents an option for the username attribute.
+	// Claim must be a singular string claim.
```
```suggestion
	// The claim's value must be a singular string.
```

enj reviewed (commented):
I have not reviewed the validation/validation tests/oidc_tests.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+	api "k8s.io/apiserver/pkg/authentication/config"
+)
+
+const (
+	authenticationConfigNilErr = "AuthenticationConfiguration can't be nil"
+	atLeastOneRequiredErrFmt   = "at least one %s is required"
+	issuerURLSchemeErr         = "issuer URL scheme must be https"
+	issuerURLRequiredErr       = "issuer URL is required"
+)
```
Unless these are used more than once, I would just inline them.  The pattern of using consts for these like in the encryption config validation is bad approach IMO.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+
+// ValidateAuthenticationConfiguration validates a given AuthenticationConfiguration.
+func ValidateAuthenticationConfiguration(c *api.AuthenticationConfiguration) field.ErrorList {
+	allErrs := field.ErrorList{}
```
```suggestion
	var allErrs field.ErrorList
```

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+		return allErrs
+	}
+
+	if len(c.JWT) == 0 {
+		allErrs = append(allErrs, field.Required(root, fmt.Sprintf(atLeastOneRequiredErrFmt, root)))
+		return allErrs
+	}
```
Hmm I guess this is okay _for now_ but as we add stuff to `AuthenticationConfiguration` in the future, this will be too strict.

aramase replied:
Agreed! Some of the validation that I added now are solely based on what we support now but I'll add a comment for reference.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	seenIssuers := sets.NewString()
+	for i, a := range c.JWT {
```
Check that you have only 1.

aramase replied:
Removed the `seenIssuers` set as we only support 1 JWT authenticator and added a comment!

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+func validateJWTAuthenticator(authenticator api.JWTAuthenticator, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if authenticator.Issuer.URL == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("issuer", "url"), issuerURLRequiredErr))
+	} else {
+		u, err := url.Parse(authenticator.Issuer.URL)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("issuer", "url"), authenticator.Issuer.URL, err.Error()))
+		} else if u.Scheme != "https" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("issuer", "url"), authenticator.Issuer.URL, issuerURLSchemeErr))
+		}
+	}
```
Split this into its own function so you can use `return` to avoid `else` statements.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+		}
+	}
+
+	if len(authenticator.Issuer.ClientIDs) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("issuer", "clientIDs"), fmt.Sprintf(atLeastOneRequiredErrFmt, fldPath.Child("issuer", "clientIDs"))))
+	}
```
Also validate that we only have 1.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+	return allErrs
+}
+
+func validateJWTAuthenticator(authenticator api.JWTAuthenticator, fldPath *field.Path) field.ErrorList {
```
You need to validate the CA bundle if non-empty.

enj replied:
Also add the strictest validation that you can think of for `ClaimValidationRules` and `ClaimMappings`.  For example, the `Username` claim mapping is required.  Specifying a `Prefix` without a `Claim` in a `PrefixedClaimOrExpression` is invalid, etc.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+func validateJWTAuthenticator(authenticator api.JWTAuthenticator, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if authenticator.Issuer.URL == "" {
```
Kube convention:

```suggestion
	if len(authenticator.Issuer.URL) == 0 {
```

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+	if authenticator.Issuer.URL == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("issuer", "url"), issuerURLRequiredErr))
+	} else {
+		u, err := url.Parse(authenticator.Issuer.URL)
```
Also validate:

- `u.User` must be `nil`
- `u.RawQuery` must be empty
- `u.Fragment` must be empty

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	if len(authenticator.Issuer.ClientIDs) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("issuer", "clientIDs"), fmt.Sprintf(atLeastOneRequiredErrFmt, fldPath.Child("issuer", "clientIDs"))))
```
Check that each client ID itself is not the empty string, and also look up the rules for valid client names in the OAuth/OIDC spec.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,88 @@
+)
+
+// ValidateAuthenticationConfiguration validates a given AuthenticationConfiguration.
+func ValidateAuthenticationConfiguration(c *api.AuthenticationConfiguration) field.ErrorList {
```
I think one of the newer API conventions is to require a max size on all lists.  To copy the authz KEP, I think we should enforce that all lists have a max of `64` items.

enj reviewed (commented):
I still have not looked at the validation or oidc_test.go

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/v1alpha1/types.go:
```diff
@@ -0,0 +1,109 @@
+	ClientIDs []string `json:"clientIDs"`
+}
+
+type ClaimValidationRule struct {
+	// claim is the name of a required claim.
+	// Same as --oidc-required-claim flag.
+	// Only string claims are supported.
+	// +optional
+	Claim string `json:"claim"`
+	// requiredValue is the value of a required claim.
+	// Same as --oidc-required-claim flag.
+	// +optional
+	RequiredValue string `json:"requiredValue"`
```
Previously this was a map, now we need to validate that the claim names do not overlap.

enj commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go:
These mechanical changes looks fine.  Are there any integration tests we want to add?

aramase replied:
Relying on the newly added integration tests to pass as we're still using the same flags and defaulting in this PR with additional validation.

aramase commented on pkg/kubeapiserver/authenticator/config.go:
```diff
@@ -153,33 +146,28 @@ func (config Config) New() (authenticator.Request, *spec.SecurityDefinitions, er
-			oidcCAContent, oidcCAErr = staticCAContentProviderFromFile("oidc-authenticator", config.OIDCCAFile)
-			if oidcCAErr != nil {
-				return nil, nil, oidcCAErr
+	if config.AuthenticationConfig != nil {
+		for _, jwtAuthenticator := range config.AuthenticationConfig.JWT {
```
I was planning to add this in validation and I see your comment with the similar suggestion: https://github.com/kubernetes/kubernetes/pull/118984#discussion_r1263743895, so going to resolve this.

aramase commented:
@enj Added the additional validations and addressed review comments in https://github.com/kubernetes/kubernetes/pull/118984/commits/a28e9453978ec8e613a8f38d6804b0feac1e6e7e. I haven't made any defaulting changes yet because we're configuring the struct based on flag first and then validating it (checking if the username claim does exist). PTAL!

enj commented:
/milestone next-candidate

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation_test.go:
```diff
@@ -0,0 +1,402 @@
+	}
+}
+
+func TestValidateCertificateAuthority(t *testing.T) {
```
Test the valid case?

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation_test.go:
```diff
@@ -0,0 +1,402 @@
+	}
+}
+
+func TestValidateClaimMappings(t *testing.T) {
```
Test the valid case?

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/v1alpha1/types.go:
```diff
@@ -0,0 +1,110 @@
+type AuthenticationConfiguration struct {
+	metav1.TypeMeta
+
+	// jwt is a list of authenticator to authenticate Kubernetes users using
```
Reminder to update this based on our conversation from yesterday.

aramase replied:
Made an update to this. LMK what you think!

aramase commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,194 @@
+	for i, clientID := range clientIDs {
+		fldPath := fldPath.Index(i)
+		if len(clientID) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath, "client ID can't be empty"))
```
Based on this comment, do we want any additional validation? As of now I only have the check to see if client ID is non-empty.

enj commented:
/milestone v1.29

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,204 @@
+// ValidateJWTAuthenticator validates a given JWTAuthenticator.
+// This is exported for use in oidc package.
+func ValidateJWTAuthenticator(authenticator api.JWTAuthenticator) field.ErrorList {
+	return validateJWTAuthenticator(authenticator, root.Index(0))
```
```suggestion
	return validateJWTAuthenticator(authenticator, field.NewPath(""))
```

The error messages from this function should not mention `jwt[0]`

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,204 @@
+	}
+	_, err := cert.NewPoolFromBytes(certificateAuthority)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, string(certificateAuthority), err.Error()))
```
```suggestion
		allErrs = append(allErrs, field.Invalid(fldPath, "<omitted>", err.Error()))
```

I suspect this error message will explode with noise if you have a CA bundle with a typo...

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/testdata/generate.sh:
Can we not use shell scripts to generate CAs?  For example, see the CSR duration test for how to generate certs via our utility funcs:

https://github.com/kubernetes/kubernetes/blob/201f8070369c15de231bbd82267a6c990f8f3d2b/test/integration/certificates/duration_test.go#L95

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation_test.go:
```diff
@@ -0,0 +1,389 @@
+		{
+			name: "authentication configuration is nil",
+			in:   nil,
+			want: "[]: Required value: AuthenticationConfiguration can't be nil",
```
The `[]: ` part looks odd?

aramase replied:
I think the more appropriate place for checking if `AuthenticationConfiguration != nil` is at the caller instead of the validate function, so I've removed that check and the corresponding test. In the feature gate PR, I'll add the nil check before calling validate.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation_test.go:
```diff
@@ -0,0 +1,389 @@
+					},
+				},
+			},
+			want: "jwt[0].issuer.url: Invalid value: \"invalid-url\": URL scheme must be https",
```
Please use raw strings to avoid all the escaping.

enj reviewed (commented):
Minor stuff.

enj commented:
/lgtm
/approve

/assign liggitt 

For API server wiring changes and API review on authentication config.

liggitt commented on pkg/kubeapiserver/authenticator/config.go:
```diff
@@ -153,33 +146,28 @@ func (config Config) New() (authenticator.Request, *spec.SecurityDefinitions, er
+	if config.AuthenticationConfig != nil {
+		for _, jwtAuthenticator := range config.AuthenticationConfig.JWT {
+			var oidcCAContent oidc.CAContentProvider
+			if jwtAuthenticator.Issuer.CertificateAuthority != nil {
```
```suggestion
			if len(jwtAuthenticator.Issuer.CertificateAuthority) > 0 {
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/authentication/config/validation/validation.go:
```diff
@@ -0,0 +1,199 @@
+		allErrs = append(allErrs, field.Required(fldPath.Child("username", "claim"), "claim name is required"))
+	}
+
+	if len(m.Groups.Prefix) != 0 && len(m.Groups.Claim) == 0 {
```
require prefix if groups claim is set, require username prefix always?

aramase commented:
@liggitt @enj pushed changes in https://github.com/kubernetes/kubernetes/pull/118984/commits/f9483a1ab18ff0f2d3e42f1fc38ad53a14983ea4. PTAL! I'll squash them back to 2 commits after the review.

enj commented on pkg/kubeapiserver/options/authentication.go:
```diff
@@ -397,16 +401,68 @@ func (o *BuiltInAuthenticationOptions) ToAuthenticationConfig() (kubeauthenticat
+			ClaimMappings: apiserver.ClaimMappings{
+				Username: apiserver.PrefixedClaimOrExpression{
+					Prefix: pointer.String(usernamePrefix),
+					Claim:  o.OIDC.UsernameClaim,
```
For CLI flags we want the `sub` defaulting logic right, and for the config path we want to make it a required config?

aramase replied:
For the CLI flag, the default is set as part of the flag

https://github.com/kubernetes/kubernetes/blob/595a57940a45f25bad4621548ee990c81779c12e/pkg/kubeapiserver/options/authentication.go#L284-L287

> config path we want to make it a required config

we want to make `claimMappings.Username` required?

aramase replied:
Made `claimMappings.Username` required.

aramase commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -597,33 +541,36 @@ func (a *Authenticator) AuthenticateToken(ctx context.Context, token string) (*a
 
-	if a.usernamePrefix != "" {
-		username = a.usernamePrefix + username
+	if a.jwtAuthenticator.ClaimMappings.Username.Prefix != nil && *a.jwtAuthenticator.ClaimMappings.Username.Prefix != "" {
```
https://github.com/kubernetes/kubernetes/issues/120049

aramase commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -158,3 +158,111 @@ type TracingConfiguration struct {
+	// claimMappings points claims of a token to be treated as user attributes.
+	// If unset, claimMappings will be defaulted to:
+	//     (1) claimMappings.username.Claim: "sub"
+	//     (2) claimMappings.username.Prefix: "" (no prefix will be prepended to the username)
```
removed it based on our discussion (xref: https://kubernetes.slack.com/archives/C04UMAUC4UA/p1692051852177559)

enj commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -158,3 +158,113 @@ type TracingConfiguration struct {
+	Username PrefixedClaimOrExpression `json:"username"`
+	// groups represents an option for the groups attribute.
+	// The claim's value must be a string or string array claim.
+	// If groups.Claim is set, the prefix is required.
```
```suggestion
	// If groups.Claim is set, the prefix must be specified (and can be the empty string).
```

enj commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -158,3 +158,113 @@ type TracingConfiguration struct {
+	// claim is the JWT claim to use.
+	// +optional
+	Claim string `json:"claim"`
+	// prefix is prepended to claim's value to prevent clashes with existing names.
+	// +required
+	Prefix *string `json:"prefix"`
```
@liggitt I assume this won't be required once we add expression field, so what should the tag be set to `optional` or `required`?

liggitt replied:
can leave it +required for now and switch to +optional once we have another field, or just make it +optional with good docs now

the tag doesn't drive any behavior and we don't have openapi for this config file API, so it doesn't *really* matter... we just don't want to be confusing

enj commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go:
```diff
@@ -157,3 +157,113 @@ type TracingConfiguration struct {
+type AuthenticationConfiguration struct {
+	metav1.TypeMeta
+
+	// jwt is a list of authenticator to authenticate Kubernetes users using
```
It is probably easier to drop the docs on the internal types since I think we will go through a few iterations of the verbiage anyway.

enj commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go:
```diff
@@ -0,0 +1,205 @@
+func validateClaimMappings(m api.ClaimMappings, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// In case of --oidc-username-claim flag, the default value is "sub".
```
Out of date comment.

enj replied:
Actually maybe not out of date, just confused me when I read it because I thought it was talking about the config based approach.  I will let you decide if you want to change anything here.

enj commented on pkg/kubeapiserver/options/authentication.go:
```diff
@@ -397,16 +401,68 @@ func (o *BuiltInAuthenticationOptions) ToAuthenticationConfig() (kubeauthenticat
-		ret.OIDCUsernameClaim = o.OIDC.UsernameClaim
-		ret.OIDCUsernamePrefix = o.OIDC.UsernamePrefix
+	if o.OIDC != nil && len(o.OIDC.IssuerURL) > 0 && len(o.OIDC.ClientID) > 0 {
+		usernamePrefix := o.OIDC.UsernamePrefix
+
+		if o.OIDC.UsernamePrefix == "" && o.OIDC.UsernameClaim != "email" {
+			// Legacy CLI flag behavior. If a usernamePrefix isn't provided, prefix all claims other than "email"
+			// with the issuerURL.
+			//
+			// See https://github.com/kubernetes/kubernetes/issues/31380
+			usernamePrefix = o.OIDC.IssuerURL + "#"
+		}
+		if o.OIDC.UsernamePrefix == "-" {
+			// Special value indicating usernames shouldn't be prefixed.
+			usernamePrefix = ""
+		}
```
Can we expand `TestBuiltInAuthenticationOptionsAddFlags` to exercise this code path with some table tests?  I want to make sure that getting to this spot via flags behaves the way we expect it to.

aramase replied:
`TestBuiltInAuthenticationOptionsAddFlags` is specifically testing `AddFlags`, so wasn't sure if that's the one we wanted to expand. I've expanded the newly added `TestToAuthenticationConfig_OIDC` to take flags as input to validate `AddFlags` and `ToAuthenticationConfig` exercising that code path.

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -158,3 +158,113 @@ type TracingConfiguration struct {
+	//
+	// In the flag based approach, the --oidc-username-claim and --oidc-username-prefix are optional. If --oidc-username-claim is not set,
+	// the default value is "sub". For the authentication config, there is no defaulting for claim or prefix. The claim and prefix must be set explicitly.
+	// For claim, if --oidc-username-claim was not set with legacy flag approach, configure userName.claim="sub" in the authentication config.
```
```suggestion
	// For claim, if --oidc-username-claim was not set with legacy flag approach, configure username.claim="sub" in the authentication config.
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -158,3 +158,113 @@ type TracingConfiguration struct {
+	// For claim, if --oidc-username-claim was not set with legacy flag approach, configure userName.claim="sub" in the authentication config.
+	// For prefix:
+	//     (1) --oidc-username-prefix="-", no prefix was added to the username. For the same behavior using authentication config,
+	//         set userName.prefix=""
```
```suggestion
	//         set username.prefix=""
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -158,3 +158,113 @@ type TracingConfiguration struct {
+	//     (1) --oidc-username-prefix="-", no prefix was added to the username. For the same behavior using authentication config,
+	//         set userName.prefix=""
+	//     (2) --oidc-username-prefix="" and  --oidc-username-claim != "email", prefix was "<value of --oidc-issuer-url>#". For the same
+	//         behavior using authentication config, set userName.prefix="<value of issuer.url>#"
```
```suggestion
	//         behavior using authentication config, set username.prefix="<value of issuer.url>#"
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -158,3 +158,113 @@ type TracingConfiguration struct {
+	//         set userName.prefix=""
+	//     (2) --oidc-username-prefix="" and  --oidc-username-claim != "email", prefix was "<value of --oidc-issuer-url>#". For the same
+	//         behavior using authentication config, set userName.prefix="<value of issuer.url>#"
+	//	   (3) --oidc-username-prefix="<value>". For the same behavior using authentication config, set userName.prefix="<value>"
```
```suggestion
	//	   (3) --oidc-username-prefix="<value>". For the same behavior using authentication config, set username.prefix="<value>"
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -158,3 +158,113 @@ type TracingConfiguration struct {
+	Username PrefixedClaimOrExpression `json:"username"`
+	// groups represents an option for the groups attribute.
+	// The claim's value must be a string or string array claim.
+	// // If groups.Claim is set, the prefix must be specified (and can be the empty string).
```
```suggestion
	// // If groups.claim is set, the prefix must be specified (and can be the empty string).
```

liggitt commented:
/approve
/lgtm
API bits lgtm.

/hold
Can fix the godoc nits in this PR or in a follow-up if you want, up to you

aramase commented on pkg/kubeapiserver/options/authentication.go:
```diff
@@ -397,16 +401,68 @@ func (o *BuiltInAuthenticationOptions) ToAuthenticationConfig() (kubeauthenticat
-		ret.OIDCIssuerURL = o.OIDC.IssuerURL
-		ret.OIDCUsernameClaim = o.OIDC.UsernameClaim
-		ret.OIDCUsernamePrefix = o.OIDC.UsernamePrefix
+	if o.OIDC != nil && len(o.OIDC.IssuerURL) > 0 && len(o.OIDC.ClientID) > 0 {
```
yes, I'm doing that as part of https://github.com/kubernetes/kubernetes/pull/119142. I'll rebase that on top of this PR now.

aramase commented:
/hold cancel

@liggitt fixed the godoc comments!

liggitt commented:
/lgtm
